### PR TITLE
dataset_utils: add mask_out_tokens to train_on_responses_only (fixes unslothai/unsloth#6695)

### PR DESCRIPTION
**Pull request by pjordanandrsn**
Implements the feature requested in unslothai/unsloth#6695: a way to keep specific tokens —
canonically the `</think>` closer — out of the loss even inside kept response spans (the
Nemotron-Ultra recipe the issue cites).

## What it does

```python
train_on_responses_only(trainer, instruction_part=..., response_part=...,
                        mask_out_tokens=["</think>"])
```

After response spans are unmasked, each listed string is re-masked to −100 wherever its token-id
sequence occurs. Matching is by tokenized id sequence **plus a leading-space variant** (covers
SentencePiece-style in-context merges); atomic added tokens like `</think>` always match exactly.
The scan is O(n) integer comparisons inside the existing per-example closure; sequences are
precomputed once outside it.

## Verification (real Qwen3-0.6B chat render, two assistant turns)

- **Baseline behavior preserved:** without the kwarg, output labels are identical to current main
  (asserted elementwise on the rendered example).
- **The motivating case:** current main trains on `</think>` (id 151668) inside the kept response;
  with `mask_out_tokens=["</think>"]` the label diff is exactly that position set to −100 — nothing
  else changes.
- **Multi-token strings:** `mask_out_tokens=["Final answer:"]` masks exactly the trainable
  positions of both in-context occurrences (matched via the bare `[19357, 4226, 25]` and
  space-prefixed `[13023, 4226, 25]` variants), all set to −100, nothing else touched.

Happy to add a unit test in your suite mirroring the above (render → label-diff assertions) if
you'd like it in-tree — say the word and I'll push it to this branch.

### CODE REVIEW

File: unsloth_zoo/dataset_utils.py [hunk 1]
```diff
@@ -323,12 +323,21 @@ def train_on_responses_only(
     return_function   = False, # Useful for iterating over lists
     num_proc          = None,
     last_response_only = False, # Train only on the last assistant turn
+    mask_out_tokens   = None,  # e.g. ["</think>"] - also mask these inside kept responses
```
[the remark this answers is not in the document]
Comment (pjordanandrsn) [reply unsloth_zoo/dataset_utils.py:682]: Fixed in 76587db — the vision-collator recursive call now passes mask_out_tokens through, so the VLM path applies the same re-masking instead of silently ignoring the kwarg. Text-path label output is regression-identical.

[the remark this answers is not in the document]
Comment (pjordanandrsn) [reply unsloth_zoo/dataset_utils.py:682]: Fixed in 76587db — the MLX wrapper signature now accepts mask_out_tokens and forwards it to the HF implementation, mirroring the rest of the API surface.

**Comment by pjordanandrsn** [comment]
Rebased onto current `main` (`33996897`) — this was sitting conflicting.

Two things worth flagging, since the diff changed slightly beyond a plain rebase:

**The conflict** was with #992, which wrapped the vision-collator branch in an `if/else`. Resolved by keeping that structure and adding `mask_out_tokens` to the nested call.

**One line is new.** #790 added a delegation to the MLX trainer's `train_on_responses_only` (`dataset_utils.py:708`) that doesn't forward `mask_out_tokens`. This PR's second commit already added that parameter to the MLX function, so a straight rebase would have left the MLX path accepting the argument and silently ignoring it — no error, just nothing masked. Added the forward there.

The two `vision_utils.py` call sites still don't take `mask_out_tokens`; those are unchanged from this PR's original scope, so I've left them alone rather than widen it here.

Verified locally on a real tokenizer: with `mask_out_tokens=["</think>"]`, the `</think>` position gets `-100` while the rest of the response span stays trainable.

**Comment by danielhanchen** [comment]
Checked this against main, where train_on_responses_only has no way to keep the thinking close tag out of the loss inside kept spans, and the default path stays byte-identical since the id-sequence scan only runs when the argument is set. Please do push the in-tree test you offered, covering the tensor and the vision collator paths, and I will get this reviewed.

**Comment by pjordanandrsn** [comment]
Pushed the test: `tests/test_train_on_responses_mask_out_tokens.py`, run in the behavioral-gates step. It covers the list path (incl. `datasets.map`), the tensor path as `UnslothVisionDataCollator.__call__` hands batches over, the vision-collator path through `train_on_responses_only(trainer, ...)`, and the MLXTrainer delegation, each with a control that leaves the argument out, plus one real `PreTrainedTokenizerFast` built in memory. No weights, no network.

Running the full gate step against main also caught `test_train_on_responses_only_forwards_last_response_only`: its keyword-only fake rejects the wrapper's new `mask_out_tokens` with a TypeError, latent since the second commit because CI never ran here. Extended the fake and asserted the keyword arrives, and rebased onto a5adec73; the PR diff is unchanged.
